### PR TITLE
Transition HBResponder to be async/await, removing HBAsyncResponder in the process

### PR DESCRIPTION
**Pull request by Joannis**
This PR provides a compatibility helper for APIs still working with ELFs, primarily the Middleware

### CODE REVIEW

File: Sources/Hummingbird/AsyncAwaitSupport/AsyncMiddleware.swift [hunk 1]
```diff
@@ -51,13 +51,13 @@ struct HBPropagateServiceContextResponder<Context: HBTracingRequestContext>: HBR
     let responder: any HBResponder<Context>
     let context: Context
```
[the remark this answers is not in the document]
Comment (Joannis) [reply Sources/Hummingbird/AsyncAwaitSupport/AsyncMiddleware.swift:53]: AsyncMiddleware comes in a different PR (#257)

File: Sources/Hummingbird/Middleware/TracingMiddleware.swift [hunk 2]
```diff
@@ -38,74 +38,74 @@ public struct HBTracingMiddleware<Context: HBTracingRequestContext & HBRemoteAdd
...
+            InstrumentationSystem.instrument.extract(request.headers, into: &serviceContext, using: HTTPHeadersExtractor())
 
-        let operationName: String = {
-            guard let endpointPath = context.endpointPath else {
-                return "HTTP \(request.method.rawValue) route not found"
-            }
-            return endpointPath
-        }()
-
-        return context.withSpan(operationName, serviceContext: serviceContext, ofKind: .server) { context, span in
```
[the remark this answers is not in the document]
Comment (Joannis) [reply Sources/Hummingbird/Middleware/TracingMiddleware.swift:51]: Isn't this applicable only after we land the HBMiddleware PR?

File: Sources/Hummingbird/Server/Responder.swift [hunk 3]
```diff
@@ -13,26 +13,28 @@
...
-    public init(callback: @escaping @Sendable (HBRequest, Context) -> EventLoopFuture<HBResponse>) {
+    public init(callback: @escaping @Sendable (HBRequest, Context) async throws -> HBResponse) {
         self.callback = callback
     }
 
-    /// Return EventLoopFuture that will be fulfilled with response to the request supplied
-    public func respond(to request: HBRequest, context: Context) -> EventLoopFuture<HBResponse> {
-        return self.callback(request, context)
+    public func respond(to request: HBRequest, context: Context) async throws -> HBResponse {
+        return try await ServiceContext.$current.withValue(context.serviceContext) {
```
[the remark this answers is not in the document]
Comment (Joannis) [reply Sources/Hummingbird/Server/Responder.swift:36]: That doesn't apply until middleware also are async right?

File: Tests/HummingbirdTests/TracingTests.swift [hunk 4]
```diff
@@ -370,8 +373,8 @@ final class TracingTests: XCTestCase {
         let app = HBApplicationBuilder(requestContext: HBTestRouterContext.self)
         app.middleware.add(SpanMiddleware())
         app.middleware.add(HBTracingMiddleware())
-        app.router.get("/") { _, context -> EventLoopFuture<HTTPResponseStatus> in
-            return context.eventLoop.scheduleTask(in: .milliseconds(2)) { return .ok }.futureResult
+        app.router.get("/") { _, context -> HTTPResponseStatus in
+            return try await context.eventLoop.scheduleTask(in: .milliseconds(2)) { return .ok }.futureResult.get()
```
[the remark this answers is not in the document]
Comment (Joannis) [reply Tests/HummingbirdTests/TracingTests.swift:381]: Is that available on all supported macOS versions?

File: Sources/Hummingbird/Server/Application+HTTPResponder.swift [hunk 5]
```diff
@@ -40,8 +40,11 @@ extension HBApplication {
                 logger: loggerWithRequestId(self.applicationContext.logger)
             )
             let httpVersion = request.version
+            
             // respond to request
-            self.responder.respond(to: request, context: context).whenComplete { result in
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/Hummingbird/Server/Application+HTTPResponder.swift:44]: We can leave this for another PR

**Review by adam-fowler — APPROVED** [review]
A good start

File: Tests/HummingbirdTests/TracingTests.swift [hunk 6]
```diff
@@ -378,8 +385,9 @@ final class TracingTests: XCTestCase {
         let router = HBRouterBuilder(context: HBTestRouterContext.self)
         router.middlewares.add(SpanMiddleware())
         router.middlewares.add(HBTracingMiddleware())
-        router.get("/") { _, context -> EventLoopFuture<HTTPResponseStatus> in
-            return context.eventLoop.scheduleTask(in: .milliseconds(2)) { return .ok }.futureResult
+        router.get("/") { _, context -> HTTPResponseStatus in
+            return .ok
+            try await Task.sleep(for: .milliseconds(2))
```
Comment (adam-fowler) [inline Tests/HummingbirdTests/TracingTests.swift:390]: Did you mean to have the sleep after the return

**Review by adam-fowler — APPROVED** [review]
One minor issue

File: Tests/HummingbirdTests/TracingTests.swift (hunk 6 above)
Comment (Joannis) [reply Tests/HummingbirdTests/TracingTests.swift:390]: Oops, that's a merge mistake

**Pull request merged**